src/pools/pool.ts

```diff
@@ -0,0 +1,26 @@
+import type { PoolEmitter } from './abstract-pool'
+
+export type ErrorHandler<Worker> = (this: Worker, e: Error) => void
+export type OnlineHandler<Worker> = (this: Worker) => void
+export type ExitHandler<Worker> = (this: Worker, code: number) => void
+
+export interface IWorker {
+  on(event: 'error', handler: ErrorHandler<this>): void
+  on(event: 'online', handler: OnlineHandler<this>): void
+  on(event: 'exit', handler: ExitHandler<this>): void
+}
+
+export interface IPool<
+  Worker extends IWorker,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Data = any,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Response = any
+> {
+  readonly workers: Worker[]
+  // nextWorker: number // Do we want to define this here?
+  readonly tasks: Map<Worker, number>
+  readonly emitter: PoolEmitter
+  destroy(): Promise<void>
+  execute(data: Data): Promise<Response>
```

Comment:
@pioardi 
I currently exposed **any** public definitions
_(excluding `nextWorker`, because I don't think that we really need to expose this one)_

But should we define `workers`, `tasks` and `emitter`?
Users could also use things like `class MyOwnPool implements AbstractPool<IWorker> {}`

Reply:
We should export only what we want the end user to use.
The AbstractPool is not intended to be exposed for me.

Reply:
I would suggest to expose it, so users can potentially create their own implementation

But I think I will remove `workers`, `tasks` and `emitter` from `IPool` because the have nothing todo with the usage of the pool, but only its values in `AbstractPool` behinde the scenes  
So if users want to have these they can use `AbstractPool` as interface instead of extending it

Reply:
So if an end user want to implement a pool would be better that implement it into our project so that is shared with everyone, right?

This is why I prefer to not expose the AbstractPool, do you agree ?

Reply:
I personally if we use the friendly way and relay on a positive biased community that are hyped to provide features to our project instead of forcing them and hiding stuff from them just because we can :slightly_smiling_face: 

But you have the last word on that, because you are the main-author of the project -> so yes: expose, or no: not expose

Reply:
Oh I would prefer not, but I don't want to have the last word just because I am the author.
So we have 1 vote for yes and 1 vote for no.

@jerome-benoit  your vote will win :)

Reply:
@jerome-benoit 

<img src="https://miro.medium.com/max/1200/1*lEWHJVVp-CzJSKhfHAfPow.jpeg" width="300px"/>

_so I can proceed to work_ :smile:

Reply:
@pioardi if @jerome-benoit doesn't see this, should we flip a coin? :laughing:  
Or should we advertise this request-for-vote more to him?
_I want to proceed_ .-.

Reply:
haha @Shinigami92  you are so fast, don't worry. Let's wait 1/2 hours and if he does not see there I will let you know :) 

You are working so much on this, thanks my friend

Reply:
I will go play Heroes of the Storm now, will come back after some rounds :)

Reply:
Three rounds HotS finished

Reply:
@ST-DDT maybe you can help with a vote here :thinking:

Reply:
I don't want to block you, if you can to proceed go with no export for the AbstractPool or wait a bit :D

And yes if @ST-DTT wants to contribute to this project I would be happy to expand the community.
I way that he/she reviewed already some PR

Reply:
He is my co-worker personal friend ^^
Okay, I will remove the export as you said and we can discuss this later

Reply:
Done https://github.com/pioardi/poolifier/pull/116/commits/de5c6d7d3bb128532ac86b684f6a3eafd56b464b

Reply:
Thanks @Shinigami92 

Your co-worker and friend is welcome on this community as each other :) Collaborator grants will be added only after a while but he can still contribute when he wants :)

Reply:
He is too fast...^^
I know it's a little late but:

- If it's internal API, then don't expose it.
- If it's a useful base class, then expose it.

In this case, I don't see anything internal in there, so I would "vote" to expose it, unless it is likely to change.
The reason for that is simple, if the base class is not exposed, then other developers won't use it to create their own implementations, which usually results in them not sharing their implementation (because it's not compatible anyway). If you put a rock on the road, then people will take another road instead of climbing over it, unless they have to.

Reply:
Hei :) So yes I understand your position( shini and you). It can make sense to expose it maybe but we will need to document really well all the internal methods and give more guide on how to create a pool then.

I will not oppose further to that, it is ok for me, we can anyway handle this later on

Reply:
> @jerome-benoit
> 
> <img alt="" width="300px" src="https://camo.githubusercontent.com/b66acaac92aa0292c70fe9f8347ded13361c2ed3f995fe0b543336a7933b09ce/68747470733a2f2f6d69726f2e6d656469756d2e636f6d2f6d61782f313230302f312a6c4557484a5656702d437a4a534b6866484166506f772e6a706567">
> 
> _so I can proceed to work_ 😄

I'm a bit busy with my paid work currently and I must have a look to the code in that PR to make up a sensible position. 

All I can say right now is that exposing to the outside world abstract class in a thing that I usually not do. I only expose the interface meant to be consumed by others implemented in the class inherited from the abstract class.

Reply:
Thx for all your points

Lets bring the discussion to #123 (you should copy your texts to there so I don't need to copy your related work in my name)